frontend/src/tests/unit/Api.test.js

```diff
@@ -0,0 +1,84 @@
+import { convertDateToStringInfo, getRemainingDate, parseJwt } from '../../utils/Api'
+import { commonInvalidObjects } from './storage/commonInvalidObjects'
+import { testEndDates } from './storage/endDates'
+import { testInvalidTokens, testProfessorTokens, testStudentTokens } from './storage/tokens'
+
+// useFakeTimers does not work in describe-scope so we apply it to all tests in the file
+// it does not change anything in them anyway
+beforeAll(() => {
+  jest.useFakeTimers().setSystemTime(new Date('2022-05-13T12:00:00.000Z'))
+})
+
+afterAll(() => {
+  jest.useRealTimers()
+})
+
+describe('parseJwt() tests', () => {
+  it.each(testStudentTokens)('student token %s parsed correctly', (token) => {
+    //when
+    const parsedToken = parseJwt(token.access_token)
+
+    //then
+    expect(parsedToken).toEqual(token.token_data)
+  })
+  it.each(testProfessorTokens)('professor token %s parsed correctly', (token) => {
+    //when
+    const parsedToken = parseJwt(token.access_token)
+
+    //then
+    expect(parsedToken).toEqual(token.token_data)
+  })
+  it.each(testInvalidTokens)('invalid token %s returns null', (token) => {
+    //when
+    const parsedToken = parseJwt(token.access_token)
+
+    //then
+    expect(parsedToken).toBeNull()
+  })
+  it.each(commonInvalidObjects)('invalid object %s returns null', (invalidObject) => {
+    //when
+    const parsedToken = parseJwt(invalidObject)
+
+    //then
+    expect(parsedToken).toBeNull()
+  })
+})
+
+describe('getRemainingDate() tests', () => {
+  it.each(testEndDates)('returns correct remaining time to date %s', (endDate) => {
+    //when
+    const remainingDate = getRemainingDate(endDate.date)
+
+    //then
+    expect(remainingDate.asDays()).toEqual(endDate.remainingDays)
+  })
+  it.each(commonInvalidObjects.filter((x) => x != undefined))(
+    'returns NaN for common invalid object %s',
+    (invalidObject) => {
+      //when
+      const remainingDate = getRemainingDate(invalidObject)
+
+      //then
+      console.log(invalidObject)
+      expect(remainingDate.asDays()).toBeNaN()
+    }
+  )
+})
```

Comment:
Ten cały test nie ma prawa działać. Przechodzi tylko dlatego, że jest źle napisany xD.
1. Jak do `console.log` dopiszesz jeszcze `remainingDate` to zobaczysz, że wcale nie jest `NaN` tylko obiektem `Duration`
2. Filtrowanie jest niepoprawne bo wyrzucasz `undefined` i liczysz na `NaN` mimo, że w funkcji tylko dla `undefined` zwracasz `NaN`
3. Z powyższych powodów linijka 15 w testowanej funkcji nigdy nie jest sprawdzana (możesz sprawdzić przez coverage)

Reply:
Moim zdaniem nie powinieneś filtrować po `undefined` tylko w tym pierwszym ifie sprawdzać czy `new Date(endDateString)` jest invalid. Jeżeli tak to `return NaN`

Reply:
Okej, generalnie przeszedłem przez to co było, sprawdziłem jak powinny wyglądać use case tych funkcji i przepisałem zarówno je jak i testy:
- z tego co daje back, data będzie zawsze postaci `number` - w millis
- refaktor funkcji żeby operowały tylko na millis
- obecnie nie mamy wpiętej daty końca do UI - zawsze zwracamy '-' (nwm czy w scope czy nie, ale mogę tam zrobić refaktor potem w innym PR)

Generalnie czemu te zmiany - bo z backu przechodzi nam data w millis po zmianach w backu, i my to castujemy na new Date() bez żadnych checków - jeśli jest tam jakiś np string, to może nam to wtedy przeparsować na jakiś bezsensowny string który będzie wyglądać na valid

Tak więc spróbuję podpiąć potem po zmianach daty końca/początku z requirements, bo obecnie nawet z tego nie korzystamy to nie zrobi nam różnicy jak działają te funkcje, a myślę że będzie warto